libraries/auth/src/main/java/org/datatransferproject/auth/OAuth1DataGenerator.java

```diff
@@ -0,0 +1,141 @@
+import org.datatransferproject.types.transfer.auth.AuthData;
+import org.datatransferproject.types.transfer.auth.TokenSecretAuthData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * General implementation of an {@link AuthDataGenerator} for OAuth1.
+ */
+public class OAuth1DataGenerator implements AuthDataGenerator {
+
+  private static final Logger logger = LoggerFactory.getLogger(OAuth1DataGenerator.class);
+  private static final String OUT_OF_BOUNDS_CALLBACK = "oob"; // TODO: is this universal?
+
+  private final OAuth1Config config;
+  private final String scope;
+  // TODO: handle dynamic updates of client ids and secrets #597
+  private final String clientId;
+  private final String clientSecret;
+  private final HttpTransport httpTransport;
+
+  OAuth1DataGenerator(OAuth1Config config, AppCredentials appCredentials,
+      HttpTransport httpTransport, String datatype, AuthMode mode) {
+    this.config = config;
+    validateConfig();
+
+    this.clientId = appCredentials.getKey();
+    this.clientSecret = appCredentials.getSecret();
+    this.httpTransport = httpTransport;
+    this.scope = mode == AuthMode.EXPORT
+        ? config.getExportScopes().get(datatype)
+        : config.getImportScopes().get(datatype);
+  }
+
+  @Override
+  public AuthFlowConfiguration generateConfiguration(String callbackBaseUrl, String id) {
+    String callback =
+        (Strings.isNullOrEmpty(callbackBaseUrl)) ? OUT_OF_BOUNDS_CALLBACK : callbackBaseUrl;
+    OAuthGetTemporaryToken tempTokenRequest = new OAuthGetTemporaryToken(
+        config.getRequestTokenUrl());
+    tempTokenRequest.callback = callback;
+    tempTokenRequest.transport = httpTransport;
+    tempTokenRequest.consumerKey = clientId;
+    tempTokenRequest.signer = config.getRequestTokenSigner(clientSecret);
+    if (config.whenAddScopes() == OAuth1Step.REQUEST_TOKEN) {
+      tempTokenRequest.set(config.getScopeParameterName(), scope);
+    }
+    TokenSecretAuthData authData;
+    try {
+      // get request token
+      OAuthCredentialsResponse tempTokenResponse = tempTokenRequest.execute();
+      authData = new TokenSecretAuthData(tempTokenResponse.token, tempTokenResponse.tokenSecret);
+    } catch (IOException e) {
+      logger.warn("Couldn't get request token {}", e.getMessage());
+      return null;
+    }
+
+    OAuthAuthorizeTemporaryTokenUrl authorizeUrl = new OAuthAuthorizeTemporaryTokenUrl(
+        config.getAuthorizationUrl());
+    authorizeUrl.temporaryToken = authData.getToken();
+    if (config.whenAddScopes() == OAuth1Step.AUTHORIZATION) {
+      authorizeUrl.set(config.getScopeParameterName(), scope);
+    }
+    String url = authorizeUrl.build();
+
+    return new AuthFlowConfiguration(url, getTokenUrl(), AuthProtocol.OAUTH_1, authData);
+  }
+
+  @Override
+  public AuthData generateAuthData(String callbackBaseUrl, String authCode, String id,
+      AuthData initialAuthData, String extra) {
```

Comment:
do we still need extra?

Reply:
I don't see it being used anywhere, but I'm reluctant to change the interface.